### ISSUE TITLE
Feature request: Optimize images in existing data uri's in CSS files

### ISSUE THREAD

**Issue by captbaritone** [parcel-bundler/parcel#8758]
# 🙋 feature request

Optimize images in existing data uri's in CSS files

## 🤔 Expected Behavior

If I have a .png (or similar) image encoded as a data uri in a CSS file, and I import that CSS file, Parcel should run image optimizations on that image file.

## 😯 Current Behavior

Parcel does not seem to run optimizations on these files.

## 💁 Possible Solution

If this is possible to implement as a plugin to Parcel, pointers to how to go about it would be appreciated as well.

## 🔦 Context

I have a [custom built postcss plugin](https://github.com/captbaritone/webamp/blob/master/packages/webamp/scripts/postcss-optimize-data-uri-pngs.js) to do this currently, and it would be nice to not need postcss after I migrate to Parcel.

## Extra

Feel free to close. I know this is quite an edge case, but I figured I'd ask, since it does seem like the type of thing Parcel might want to do.

**Comment by mischnic** [comment]
This really boils down to extracting inline data-url assets in transformers/resolver. Just like we do  for inline `<script>`s in HTML.
And same goes for `import "data:text/javascript,console.log();"` as well.

Might be nice to have this logic in the resolver so it works for every transformer, but not sure if that's feasible.

**Comment by devongovett** [comment]
I think it should be possible with a resolver plugin already, but we might need to change the type here to `string | Buffer` instead of just `string`. That way the resolver could decode the data url and return it. https://github.com/parcel-bundler/parcel/blob/v2/packages/core/types/index.js#L1519

**Comment by captbaritone** [comment]
> I think it should be possible with a resolver plugin already, but we might need to change the type here to `string | Buffer` instead of just `string`. That way the resolver could decode the data url and return it. https://github.com/parcel-bundler/parcel/blob/v2/packages/core/types/index.js#L1519

I decided to give this a try, and if I return `code` as a Buffer, Parcel throws an error here:

https://github.com/parcel-bundler/parcel/blob/2172672d175f4667c5d63afc16a90e52cf255c69/packages/core/core/src/summarizeRequest.js#L59

With a rather unhelpful error message:
<img width="955" alt="Screenshot 2023-01-12 at 8 31 58 PM" src="https://user-images.githubusercontent.com/162735/212238000-601bdc1d-a37b-4c9f-8a18-37629a7748ed.png">

Additionally, it looks like I have to supply a filePath in order for the module to resolve, and I'm not sure what I should use for a file that does not exist on disk.

**Comment by devongovett** [comment]
Yeah we'd need to change the implementation to accept buffers as a return value from a resolver instead of just strings. The filePath is required mainly for code which might have dependencies (e.g. imports) so we know where to resolve them from. That's not really the case for images I guess but probably a lot of things depend on it being there...
